src/simularium/VisGeometry/GeometryStore.ts

```diff
@@ -0,0 +1,352 @@
+import { forEach } from "lodash";
+import { OBJLoader } from "three/examples/jsm/loaders/OBJLoader";
+import jsLogger, { ILogger, ILogLevel } from "js-logger";
+import { BufferGeometry, Object3D, Mesh } from "three";
+
+import { checkAndSanitizePath } from "../../util";
+import PDBModel from "../PDBModel";
+import { InstancedMesh } from "../rendering/InstancedMesh";
+import VisAgent from "../VisAgent";
+import TaskQueue from "../worker/TaskQueue";
+import { AgentTypeVisData } from "../types";
+
+import {
+    AgentGeometry,
+    GeometryDisplayType,
+    GeometryStoreLoadResponse,
+    MeshGeometry,
+    MeshLoadRequest,
+} from "./types";
+
+export const DEFAULT_MESH_NAME = "SPHERE";
+
+type Registry = Map<string, AgentGeometry>;
+
+class GeometryStore {
+    private _geoLoadAttempted: Map<string, boolean>;
+    private _cachedAssets: Map<string, string>;
+    private _registry: Registry;
+    public mlogger: ILogger;
+
+    constructor(loggerLevel?: ILogLevel) {
+        this._geoLoadAttempted = new Map<string, boolean>();
+        this._cachedAssets = new Map<string, string>();
+        this._registry = new Map<string, AgentGeometry>();
+        this.mlogger = jsLogger.get("geometry-store");
+        if (loggerLevel) {
+            this.mlogger.setLevel(loggerLevel);
+        }
+    }
+
+    public init(): void {
+        this._geoLoadAttempted.clear();
+        this._registry.clear();
+        this._registry.set(DEFAULT_MESH_NAME, {
+            displayType: GeometryDisplayType.SPHERE,
+            geometry: {
+                mesh: new Mesh(VisAgent.sphereGeometry),
+                cancelled: false,
+                instances: new InstancedMesh(
+                    VisAgent.sphereGeometry,
+                    DEFAULT_MESH_NAME,
+                    1
+                ),
+            },
+        });
+    }
+
+    public get registry(): Registry {
+        return this._registry;
+    }
+
+    public forEachMesh(iteratee: (geo: MeshGeometry) => void): void {
+        this._registry.forEach((value) => {
+            const { displayType } = value;
+            if (displayType !== GeometryDisplayType.PDB) {
+                const agentGeo = value as MeshGeometry;
+                iteratee(agentGeo);
+            }
+        });
+    }
+
+    public getGeoForAgentType(entryName: string): AgentGeometry | null {
+        const geo = this._registry.get(entryName);
+        return geo || null;
+    }
+
+    public cancelAll(): void {
+        // note that this leaves cancelled things in the registries.
+        // This should be called before the registries are cleared and probably
+        // only makes sense to do if they are indeed about to be cleared.
+
+        // don't process any queued requests
+        TaskQueue.stopAll();
+        // signal to cancel any pending geometries
+        this._registry.forEach((value) => {
+            value.geometry.cancelled = true;
+        });
```

Comment:
dan, shouldn't this clear out `_geoLoadAttempted` too?

Reply:
I don't think it really matters.  The main point here is that any async processing is either shut down (the task queue) or when results arrive they can be discarded (the cancelled=true).